js/hw_2_js_functions_state/src/game_of_snake/game_of_snake.js

```diff
@@ -1,3 +1,453 @@
 import { createEngine, ENGINE_INITIALIZE_SIGNAL } from "../engine/engine";
 
-// Good luck!
+const meshSize = 625;
+
+const initialState = {
+  currentScore: 0,
+  gameWon: false,
+  gameLost: false,
+  shouldRestart: false,
+  pause: false,
+  time: 0,
+  snakeBody: [
+    [10, 10],
+    [10, 11],
+    [10, 12],
+    [10, 13],
+    [10, 14],
+  ],
+  food: [1 ,1],
+  direction: 'down',
+};
+
+const engine = createEngine();
+
+engine.addSignalReducer(ENGINE_INITIALIZE_SIGNAL, () => {
+  return initialState;
+});
+
+engine.addSignalReducer("restartGame", () => {
+  return initialState;
+});
+
+engine.addSideEffect({
+  onlyWhen: ifScoreChanged,
+  effect: (_, emit) => {
+    emit("createFood");
+    emit("snakeGrowUp");
```

Comment:
For multiple reasons, it is better to handle "snake growth" in the same place where the score incremented - in `snakeMove` reducer. "food creation" can and should also be a part of `snakeMove` reducer:

```
engine.addSignalReducer('snakeMove', (state) => {
   const nextHead = getNextHead(state.snake, state.direction)
   const isGameLost = getIsHeadOutsideGameBoard(nextHead, boardWidth)
   const hasEatenFood = getIsFoodThere(nextHead, snake.food)
   const nextSnake = growSnake(hasEatenFood, state.snake, nextHead)
   const nextScore = hasEatenFood ? state.score + foodBonus : state.food;
   const nextFood = hasEatenFood ? generateNewFood(state.food, state.snake, boardWidth) : state.food

   return {
       ...state,
       score: nextScore,
       gameLost: isGameLost,
       snake: nextSnake,
       food: nextFood
   }
})
```